<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>FAQ · MendelIHT</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">MendelIHT</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../getting_started/">Getting Started</a></li><li><a class="tocitem" href="../examples/">Examples</a></li><li><a class="tocitem" href="../math/">Mathematical Details</a></li><li><a class="tocitem" href="../contributing/">Contributing</a></li><li class="is-active"><a class="tocitem" href>FAQ</a><ul class="internal"><li><a class="tocitem" href="#First-time-Performance"><span>First-time Performance</span></a></li><li><a class="tocitem" href="#Memory-requirement?"><span>Memory requirement?</span></a></li><li><a class="tocitem" href="#How-to-run-code-in-parallel?"><span>How to run code in parallel?</span></a></li><li><a class="tocitem" href="#Phenotype-quality-control"><span>Phenotype quality control</span></a></li><li><a class="tocitem" href="#When-to-standardize-phenotypes?"><span>When to standardize phenotypes?</span></a></li><li><a class="tocitem" href="#When-to-standardize-covariates?"><span>When to standardize covariates?</span></a></li><li><a class="tocitem" href="#How-to-enforce-sparsity-on-non-genetic-covariates?"><span>How to enforce sparsity on non-genetic covariates?</span></a></li><li><a class="tocitem" href="#Missing-data?"><span>Missing data?</span></a></li><li><a class="tocitem" href="#Keyword-arguments?"><span>Keyword arguments?</span></a></li><li><a class="tocitem" href="#Will-IHT-work-on-sequence/imputed-data?"><span>Will IHT work on sequence/imputed data?</span></a></li></ul></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>FAQ</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>FAQ</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/MendelIHT.jl/blob/master/docs/src/man/FAQ.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="FAQ"><a class="docs-heading-anchor" href="#FAQ">FAQ</a><a id="FAQ-1"></a><a class="docs-heading-anchor-permalink" href="#FAQ" title="Permalink"></a></h1><p>If you do not find your problem here, or the provided solution does not solve your problem, please file an issue on <a href="https://github.com/OpenMendel/MendelIHT.jl/issues">GitHub</a>. </p><h2 id="First-time-Performance"><a class="docs-heading-anchor" href="#First-time-Performance">First-time Performance</a><a id="First-time-Performance-1"></a><a class="docs-heading-anchor-permalink" href="#First-time-Performance" title="Permalink"></a></h2><p>In a fresh Julia session, the first time any function gets called will take a <em>long</em> time because the code has to be compiled on the spot. For instance, compare</p><pre><code class="language-julia hljs">@time using MendelIHT</code></pre><pre><code class="nohighlight hljs">140.065958 seconds (179.19 M allocations: 8.491 GiB, 2.39% gc time, 84.11% compilation time)</code></pre><pre><code class="language-julia hljs">@time using MendelIHT</code></pre><pre><code class="nohighlight hljs">  0.000156 seconds (161 allocations: 12.859 KiB)</code></pre><p>Fortunately, for large problems, compilation time becomes negligible. </p><h2 id="Memory-requirement?"><a class="docs-heading-anchor" href="#Memory-requirement?">Memory requirement?</a><a id="Memory-requirement?-1"></a><a class="docs-heading-anchor-permalink" href="#Memory-requirement?" title="Permalink"></a></h2><p>For binary PLINK files, <code>MendelIHT.jl</code> uses <a href="https://openmendel.github.io/SnpArrays.jl/latest/#Linear-Algebra">SnpArrays.jl&#39;s SnpLinAlg</a> for linear algebra. This data structure uses memory mapping and computes directly on raw genotype file, described in our <a href="https://www.biorxiv.org/content/10.1101/2021.08.04.455145v2.abstract">multivariate paper</a>. As such, it requires roughly <span>$2np$</span> bits of <em>virtual memory</em> and much less physical memory to store. For UK biobank with 500k samples and 500k SNPs, this is roughly 62GB of virtual memory and a couple of GB of RAM. Thus, for binary PLINK files, we can easily fit IHT on large dataset, including those that are too large to fit in RAM.</p><p>In addition to storing the above matrix in memory, IHT also enables <code>memory_efficient=true</code> by default. If <code>memory_efficient=false</code>, IHT will additionally store a <span>$n \times k$</span> matrix in double precision, where <span>$k$</span> is the sparsity level. This require <span>$64nk$</span> bits of RAM. For cross validation routines that test multiple different <span>$k_1, ..., k_q$</span> values, <span>$t$</span> of them must co-exist in memory where <span>$t$</span> is the number of threads. </p><p>For BGEN and VCF files, <code>MendelIHT.jl</code> imports genotypes into double precision matrices which require <span>$64np$</span> bits of RAM. For 500k samples and 500k SNPs, this requires 2 TB of RAM. Thus, IHT does not work on extremely large VCF and BGEN files. If you have these large VCF/BGEN files, one can convert it to binary PLINK format before running IHT.</p><h2 id="How-to-run-code-in-parallel?"><a class="docs-heading-anchor" href="#How-to-run-code-in-parallel?">How to run code in parallel?</a><a id="How-to-run-code-in-parallel?-1"></a><a class="docs-heading-anchor-permalink" href="#How-to-run-code-in-parallel?" title="Permalink"></a></h2><p>If Julia is started with multiple threads (e.g. <code>julia --threads 4</code>), <code>MendelIHT.jl</code> will automatically run your code in parallel. </p><ul><li><a href="https://docs.julialang.org/en/v1/manual/multi-threading/#Starting-Julia-with-multiple-threads">How to start Julia with multiple threads</a>.</li><li>Execute <code>Threads.nthreads()</code> within Julia to check if multiple thread is enabled</li></ul><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>When Julia is started with multiple threads, make sure to set the number of BLAS threads to 1 (i.e. <code>using LinearAlgebra; BLAS.set_num_threads(1)</code>. This avoids oversubscription. </p></div></div><h2 id="Phenotype-quality-control"><a class="docs-heading-anchor" href="#Phenotype-quality-control">Phenotype quality control</a><a id="Phenotype-quality-control-1"></a><a class="docs-heading-anchor-permalink" href="#Phenotype-quality-control" title="Permalink"></a></h2><p>Our software assumes phenotypes have been properly quality controlled. For instance</p><ul><li>There no are missing values (except possibly for Gaussian traits where we impute them with the mean). </li><li>when running sparse linear regression, phenotypes should be approximately Gaussian</li><li>when running Poisson regression, phenotypes should have approximately equal mean and variance</li><li>when running multivariate traits, phenotypes are normalized</li></ul><p>...etc.</p><p>Execute your judicious judgement!</p><h2 id="When-to-standardize-phenotypes?"><a class="docs-heading-anchor" href="#When-to-standardize-phenotypes?">When to standardize phenotypes?</a><a id="When-to-standardize-phenotypes?-1"></a><a class="docs-heading-anchor-permalink" href="#When-to-standardize-phenotypes?" title="Permalink"></a></h2><p>Only multivariate Guassian traits should be standardize to mean 0 variance 1. This ensures that mean squared error in cross-validation among traits are comparable, so the tuning process is driven by all traits.</p><p>For single trait analysis, standardization is not necessary.</p><h2 id="When-to-standardize-covariates?"><a class="docs-heading-anchor" href="#When-to-standardize-covariates?">When to standardize covariates?</a><a id="When-to-standardize-covariates?-1"></a><a class="docs-heading-anchor-permalink" href="#When-to-standardize-covariates?" title="Permalink"></a></h2><p><strong>Always</strong> standardize your covariates (genetic and non-genetic) to mean 0 variance 1. This ensures sparsity is enforced equally on all predictors. </p><p>For binary PLINK files (.bed/.bim/.fam) standardization is automatic. When using wrapper functions <a href="https://openmendel.github.io/MendelIHT.jl/latest/man/api/#MendelIHT.cross_validate">cross_validate()</a> and <a href="https://openmendel.github.io/MendelIHT.jl/latest/man/api/#MendelIHT.iht">iht()</a>, non-genetic covariate will also be automatically standardized. However using internal functions <a href="https://openmendel.github.io/MendelIHT.jl/latest/man/api/#MendelIHT.fit_iht">fit_iht()</a> and <a href="https://openmendel.github.io/MendelIHT.jl/latest/man/api/#MendelIHT.cv_iht">cv_iht()</a> bypasses standardization and is generally recommended only if wrapper functions do not work for your purposes. </p><h2 id="How-to-enforce-sparsity-on-non-genetic-covariates?"><a class="docs-heading-anchor" href="#How-to-enforce-sparsity-on-non-genetic-covariates?">How to enforce sparsity on non-genetic covariates?</a><a id="How-to-enforce-sparsity-on-non-genetic-covariates?-1"></a><a class="docs-heading-anchor-permalink" href="#How-to-enforce-sparsity-on-non-genetic-covariates?" title="Permalink"></a></h2><p>The <code>zkeep</code> parameter will allow non-genetic covariates to be subject to selection. Say you have 5 covariates, and you want to always keep the first 3 in the model but possibly set the last 2 to zero. You can do</p><pre><code class="language-julia hljs">zkeep = trues(5)
zkeep[4:5] .= false
zkeep

# now input zkeep as keyword argument for the wrapper or core functions, e.g. 
# iht(plinkfile, k, d, zkeep=zkeep)</code></pre><pre><code class="nohighlight hljs">5-element BitVector:
 1
 1
 1
 0
 0</code></pre><p>Note <code>zkeep</code> is a <code>BitVector</code> and not a <code>Vector{Bool}</code>. </p><h2 id="Missing-data?"><a class="docs-heading-anchor" href="#Missing-data?">Missing data?</a><a id="Missing-data?-1"></a><a class="docs-heading-anchor-permalink" href="#Missing-data?" title="Permalink"></a></h2><p>In general, any sample or covariate with large proportion of missing (e.g. &gt;10%) should be excluded. But our software does have a few built-in mechanisms for handling them.</p><p><strong>Phenotypes:</strong> Gaussian phenotypes can be internally imputed with the mean. Binary/count phenotypes cannot be imputed.</p><p><strong>Genotypes:</strong> All genotypes can be imputed with the mean. This is the default when using wrapper functions <a href="https://openmendel.github.io/MendelIHT.jl/latest/man/api/#MendelIHT.iht">iht()</a> and <a href="https://openmendel.github.io/MendelIHT.jl/latest/man/api/#MendelIHT.cross_validate">cross_validate()</a></p><p><strong>Nongenetic covariates</strong>: These cannot be imputed. Please impute them before running IHT.</p><h2 id="Keyword-arguments?"><a class="docs-heading-anchor" href="#Keyword-arguments?">Keyword arguments?</a><a id="Keyword-arguments?-1"></a><a class="docs-heading-anchor-permalink" href="#Keyword-arguments?" title="Permalink"></a></h2><p>Julia supports 2 types of &quot;optional&quot; arguments. Optional arguments specified before semicolon <code>;</code> can be directly inputted. Optional arguments specified after semicolon <code>;</code> needs to be explicitly inputted as <code>varname = x</code>. For instance, </p><pre><code class="language-julia hljs">function add(a::Int, b::Int=1; c::Int=2)
    return a + b + c 
end
@show add(0)             # 0 + b + c using default value for b, c
@show add(0, 3)          # 0 + b + c using b = 3 and default value for c
@show add(0, 5, c=10);   # 0 + b + c using b = 5 and c = 10</code></pre><pre><code class="nohighlight hljs">add(0) = 3
add(0, 3) = 5
add(0, 5, c = 10) = 15</code></pre><h2 id="Will-IHT-work-on-sequence/imputed-data?"><a class="docs-heading-anchor" href="#Will-IHT-work-on-sequence/imputed-data?">Will IHT work on sequence/imputed data?</a><a id="Will-IHT-work-on-sequence/imputed-data?-1"></a><a class="docs-heading-anchor-permalink" href="#Will-IHT-work-on-sequence/imputed-data?" title="Permalink"></a></h2><p>If someone can test this out and tell us, that would be extremely helpful.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../contributing/">« Contributing</a><a class="docs-footer-nextpage" href="../api/">API »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Tuesday 14 March 2023 02:24">Tuesday 14 March 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
