<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · MendelIHT</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">MendelIHT</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../getting_started/">Getting Started</a></li><li><a class="tocitem" href="../examples/">Examples</a></li><li><a class="tocitem" href="../math/">Mathematical Details</a></li><li><a class="tocitem" href="../contributing/">Contributing</a></li><li><a class="tocitem" href="../FAQ/">FAQ</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Wrapper-Functions"><span>Wrapper Functions</span></a></li><li><a class="tocitem" href="#Core-Functions"><span>Core Functions</span></a></li><li><a class="tocitem" href="#Specifying-Groups-and-Weights"><span>Specifying Groups and Weights</span></a></li><li><a class="tocitem" href="#Simulation-Utilities"><span>Simulation Utilities</span></a></li><li><a class="tocitem" href="#Other-Useful-Functions"><span>Other Useful Functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/MendelIHT.jl/blob/master/docs/src/man/api.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><p>Here is a list of available function calls. A detailed description can be found below. </p><ul><li><a href="#MendelIHT.cross_validate"><code>MendelIHT.cross_validate</code></a></li><li><a href="#MendelIHT.cv_iht"><code>MendelIHT.cv_iht</code></a></li><li><a href="#MendelIHT.fit_iht"><code>MendelIHT.fit_iht</code></a></li><li><a href="#MendelIHT.iht"><code>MendelIHT.iht</code></a></li><li><a href="#MendelIHT.iht_run_many_models"><code>MendelIHT.iht_run_many_models</code></a></li><li><a href="#MendelIHT.make_bim_fam_files"><code>MendelIHT.make_bim_fam_files</code></a></li><li><a href="#MendelIHT.pve"><code>MendelIHT.pve</code></a></li><li><a href="#MendelIHT.simulate_correlated_snparray"><code>MendelIHT.simulate_correlated_snparray</code></a></li><li><a href="#MendelIHT.simulate_random_response"><code>MendelIHT.simulate_random_response</code></a></li><li><a href="#MendelIHT.simulate_random_snparray"><code>MendelIHT.simulate_random_snparray</code></a></li></ul><h2 id="Wrapper-Functions"><a class="docs-heading-anchor" href="#Wrapper-Functions">Wrapper Functions</a><a id="Wrapper-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Wrapper-Functions" title="Permalink"></a></h2><p>Most users will use the following wrapper functions. Users specify location of PLINK files and possibly the phenotype/covariate files. These functions will soon be updated to support VCF and BGEN formats.</p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.iht" href="#MendelIHT.iht"><code>MendelIHT.iht</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">iht(filename, k, d, phenotypes=6, covariates=&quot;&quot;, summaryfile=&quot;iht.summary.txt&quot;,
    betafile=&quot;iht.beta.txt&quot;, kwargs...)</code></pre><p>Runs IHT with sparsity level <code>k</code>. </p><p><strong>Arguments</strong></p><ul><li><code>filename</code>: A <code>String</code> for VCF, binary PLINK, or BGEN file. VCF files should end   in <code>.vcf</code> or <code>.vcf.gz</code>. Binary PLINK files should exclude <code>.bim/.bed/.fam</code>   trailings but the trio must all be present in the same directory. BGEN files   should end in <code>.bgen</code>.</li><li><code>k</code>: An <code>Int</code> for sparsity parameter = number of none-zero coefficients</li><li><code>d</code>: Distribution of phenotypes. Specify <code>Normal</code> for quantitative traits,   <code>Bernoulli</code> for binary traits, <code>Poisson</code> or <code>NegativeBinomial</code> for   count traits, and <code>MvNormal</code> for multiple quantitative traits. </li></ul><p><strong>Optional Arguments</strong></p><ul><li><code>phenotypes</code>: Phenotype file name (<code>String</code>), an integer, or vector of integer. Integer(s)   coresponds to the column(s) of PLINK&#39;s <code>.fam</code> file that stores phenotypes (default <code>phenotypes=6</code>).    Enter multiple integers for multivariate analysis (e.g. <code>phenotypes=[6, 7]</code>).   We recognize missing phenotypes as <code>NA</code> or <code>-9</code>. For quantitative traits   (univariate or multivariate), missing phenotypes are imputed with the mean. Binary   and count phenotypes cannot be imputed. Phenotype files are read using <code>readdlm</code> function   in Julia base. We require each subject&#39;s phenotype to occupy a different row, and for    multivariate analysis, each phenotype is comma separated. The phenotype file   should not include a header line. Each row should be listed in the same order as in   the PLINK and (for multivariate analysis) be comma separated. </li><li><code>covariates</code>: Covariate file name. Default <code>covariates=&quot;&quot;</code> (in which case an intercept   term will be automatically included). If <code>covariates</code> file specified, it will be    read using <code>readdlm</code> function in Julia base. We require the covariate file to be   comma separated, and not include a header line. Each row should be listed in the   same order as in the PLINK. The first column should be all 1s to indicate an   intercept. All other columns not specified in <code>exclude_std_idx</code> will be standardized   to mean 0 variance 1. </li><li><code>summaryfile</code>: Output file name for saving IHT&#39;s summary statistics. Default   <code>summaryfile=&quot;iht.summary.txt&quot;</code>.</li><li><code>betafile</code>: Output file name for saving IHT&#39;s estimated genotype effect sizes.    Default <code>betafile=&quot;iht.beta.txt&quot;</code>. </li><li><code>covariancefile</code>: Output file name for saving IHT&#39;s estimated trait covariance   matrix for multivariate analysis. Default <code>covariancefile=&quot;iht.cov.txt&quot;</code>. </li><li><code>exclude_std_idx</code>: Indices of non-genetic covariates that should be excluded from   standardization. </li><li><code>dosage</code>: Currently only guaranteed to work for VCF files. If <code>true</code>, will read   genotypes dosages (i.e. <code>X[i, j] ∈ [0, 2]</code> before standardizing)</li><li>All optional arguments available in <a href="#MendelIHT.fit_iht"><code>fit_iht</code></a></li></ul><p><strong>Output</strong></p><p>A fitted <code>IHTResult</code> (univariate analysis) or <code>mIHTResult</code> (multivariate analysis). In addition, a human-readable text file is also outputted with file name specified with <code>betafile</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/wrapper.jl#L1-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.cross_validate" href="#MendelIHT.cross_validate"><code>MendelIHT.cross_validate</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">cross_validate(filename, d, path=1:20, phenotypes=6, covariates=&quot;&quot;, 
    cv_summaryfile=&quot;cviht.summary.txt&quot;, q=5, kwargs...)</code></pre><p>Runs cross-validation to determinal optimal sparsity level <code>k</code>. Different sparsity levels are specified in <code>path</code>. </p><p><strong>Arguments</strong></p><ul><li><code>filename</code>: A <code>String</code> for VCF, binary PLINK, or BGEN file. VCF files should end   in <code>.vcf</code> or <code>.vcf.gz</code>. Binary PLINK files should exclude <code>.bim/.bed/.fam</code>   trailings but the trio must all be present in the same directory. BGEN files   should end in <code>.bgen</code>.</li><li><code>d</code>: Distribution of phenotypes. Specify <code>Normal</code> for quantitative traits,   <code>Bernoulli</code> for binary traits, <code>Poisson</code> or <code>NegativeBinomial</code> for   count traits, and <code>MvNormal</code> for multiple quantitative traits. </li></ul><p><strong>Optional Arguments</strong></p><ul><li><code>path</code>: Different values of <code>k</code> that should be tested. One can input a vector of    <code>Int</code> (e.g. <code>path=[5, 10, 15, 20]</code>) or a range (default <code>path=1:20</code>).</li><li><code>phenotypes</code>: Phenotype file name (<code>String</code>), an integer, or vector of integer. Integer(s)   coresponds to the column(s) of <code>.fam</code> file that stores phenotypes (default 6).    We recognize missing phenotypes as <code>NA</code> or <code>-9</code>. For quantitative traits   (univariate or multivariate), missing phenotypes are imputed with the mean. Binary   and count phenotypes cannot be imputed. Phenotype files are read using <code>readdlm</code> function   in Julia base. We require each subject&#39;s phenotype to occupy a different row, and for    multivariate analysis, each phenotype is comma separated. The phenotype file   should not include a header line. Each row should be listed in the same order as in   the PLINK. </li><li><code>covariates</code>: Covariate file name. Default <code>covariates=&quot;&quot;</code> (in which case an intercept   term will be automatically included). If <code>covariates</code> file specified, it will be    read using <code>readdlm</code> function in Julia base. We require the covariate file to be   comma separated, and not include a header line. Each row should be listed in the   same order as in the PLINK. The first column should be all 1s to indicate an   intercept. All other columns not specified in <code>exclude_std_idx</code> will be standardized   to mean 0 variance 1</li><li><code>cv_summaryfile</code>: Output file name for saving IHT&#39;s cross validation summary statistics.   Default <code>cv_summaryfile=&quot;cviht.summary.txt&quot;</code>.</li><li><code>q</code>: Number of cross validation folds. Larger means more accurate and more computationally   intensive. Should be larger 2 and smaller than 10. Default <code>q=5</code>. </li><li><code>dosage</code>: Currently only guaranteed to work for VCF files. If <code>true</code>, will read   genotypes dosages (i.e. <code>X[i, j] ∈ [0, 2]</code> before standardizing)</li><li>All optional arguments available in <a href="#MendelIHT.cv_iht"><code>cv_iht</code></a></li></ul><p><strong>Output</strong></p><p>A vector <code>mse</code> of cross-validated mean-squared errors (technically deviance residuals), where <code>mse[i]</code> is the error of sparsity level <code>path[i]</code>. In addition, a human-readable text file is also outputted with file name specified with <code>cv_summaryfile</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/wrapper.jl#L253-L300">source</a></section></article><h2 id="Core-Functions"><a class="docs-heading-anchor" href="#Core-Functions">Core Functions</a><a id="Core-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Functions" title="Permalink"></a></h2><p>Users can also use the <code>fit_iht</code> and <code>cv_iht</code> functions directly. One must import genotypes via <a href="https://github.com/OpenMendel/SnpArrays.jl">SnpArrays.jl</a> (and use <code>SnpLinAlg</code> type for <code>x</code> argument) or <a href="https://github.com/OpenMendel/VCFTools.jl">VCFTools.jl</a>. Phenotypes/covariates must also be imported using Julia&#39;s standard routine (typically with <code>Base.readdlm</code>). </p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.fit_iht" href="#MendelIHT.fit_iht"><code>MendelIHT.fit_iht</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_iht(y, x, z; k=10, d = Normal(), l=IdentityLink(),...)</code></pre><p>Fits a model on design matrix (genotype data) <code>x</code>, response (phenotype) <code>y</code>,  and non-genetic covariates <code>z</code> on a specific sparsity parameter <code>k</code>. Only predictors in  <code>x</code> will be subject to sparsity constraint, unless <code>zkeep</code> keyword is specified. </p><p><strong>Arguments:</strong></p><ul><li><code>y</code>: Phenotype vector or matrix. Should be an <code>Array{T, 1}</code> (single traits) or   <code>Array{T, 2}</code> (multivariate Gaussian traits). For multivariate traits, each    column of <code>y</code> should be a sample. </li><li><code>x</code>: Genotype matrix (an <code>Array{T, 2}</code> or <code>SnpLinAlg</code>). For univariate   analysis, samples are rows of <code>x</code>. For multivariate analysis, samples are   columns of <code>x</code> (i.e. input <code>Transpose(x)</code> for <code>SnpLinAlg</code>)</li><li><code>z</code>: Matrix of non-genetic covariates of type <code>Array{T, 2}</code> or <code>Array{T, 1}</code>.   For univariate analysis, sample covariates are rows of <code>z</code>. For multivariate   analysis, sample covariates are columns of <code>z</code>. If this is not specified, an   intercept term will be included automatically. If <code>z</code> is specified, make sure   the first column (row) is all 1s to represent the intercept. </li></ul><p><strong>Optional Arguments:</strong></p><ul><li><code>k</code>: Number of non-zero predictors. Can be a constant or a vector (for group IHT). </li><li><code>J</code>: The number of maximum groups (set as 1 if no group infomation available)</li><li><code>d</code>: Distribution of phenotypes. Specify <code>Normal()</code> for quantitative traits,   <code>Bernoulli()</code> for binary traits, <code>Poisson()</code> or <code>NegativeBinomial()</code> for   count traits, and <code>MvNormal()</code> for multiple quantitative traits. </li><li><code>l</code>: A link function. The recommended link functions are <code>l=IdentityLink()</code> for   quantitative traits, <code>l=LogitLink()</code> for binary traits, <code>l=LogLink()</code> for Poisson   distribution, and <code>l=Loglink()</code> for NegativeBinomial distribution. For multivariate   analysis, the choice of link does not matter. </li><li><code>group</code>: vector storing (non-overlapping) group membership</li><li><code>weight</code>: vector storing vector of weights containing prior knowledge on each SNP</li><li><code>zkeep</code>: BitVector determining whether non-genetic covariates in <code>z</code> will be subject    to sparsity constraint. <code>zkeep[i] = true</code> means covariate <code>i</code> will NOT be projected.   Note covariates forced in the model are not subject to sparsity constraint <code>k</code>. </li><li><code>est_r</code>: Symbol (<code>:MM</code>, <code>:Newton</code> or <code>:None</code>) to estimate nuisance parameters for negative binomial regression</li><li><code>use_maf</code>: boolean indicating whether we want to scale projection with minor allele frequencies (see paper)</li><li><code>debias</code>: boolean indicating whether we debias at each iteration</li><li><code>verbose</code>: boolean indicating whether we want to print intermediate results</li><li><code>tol</code>: used to track convergence</li><li><code>max_iter</code>: is the maximum IHT iteration for a model to converge. Defaults to 200, or 100 for cross validation</li><li><code>min_iter</code>: is the minimum IHT iteration before checking for convergence. Defaults to 5.</li><li><code>max_step</code>: is the maximum number of backtracking per IHT iteration. Defaults 3</li><li><code>io</code>: An <code>IO</code> object for displaying intermediate results. Default <code>stdout</code>.</li><li><code>init_beta</code>: Whether to initialize beta values to univariate regression values.    Currently only Gaussian traits can be initialized. Default <code>false</code>. </li><li><code>memory_efficient</code>: If <code>true,</code> it will cause ~1.1 times slow down but one only   needs to store the genotype matrix (requiring 2np bits for PLINK binary files   and <code>8np</code> bytes for other formats). If <code>memory_efficient=false</code>, one also need   to store a <code>8nk</code> byte matrix where <code>k</code> is the sparsity levels. </li></ul><p><strong>Output</strong></p><ul><li>An <code>IHTResult</code> (for single-trait analysis) or <code>mIHTResult</code> (for multivariate analysis).</li></ul><p><strong>Group IHT</strong></p><p>If <code>k</code> is a constant, then each group will have the same sparsity level. To run doubly  sparse IHT with varying group sparsities, construct <code>k</code> to be a vector where <code>k[i]</code> indicates the max number of predictors for group <code>i</code>. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/fit.jl#L1-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.cv_iht" href="#MendelIHT.cv_iht"><code>MendelIHT.cv_iht</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">cv_iht(y, x, z; path=1:20, q=5, d=Normal(), l=IdentityLink(), ...)</code></pre><p>For each model specified in <code>path</code>, performs <code>q</code>-fold cross validation and  returns the (averaged) deviance residuals. The purpose of this function is to find the best sparsity level <code>k</code>, obtained from selecting the model with the minimum out-of-sample error. Note sparsity is enforced on <code>x</code> only, unless <code>zkeep</code> keyword is specified. </p><p>To check if multithreading is enabled, check output of <code>Threads.nthreads()</code>.</p><p><strong>Arguments:</strong></p><ul><li><code>y</code>: Phenotype vector or matrix. Should be an <code>Array{T, 1}</code> (single traits) or   <code>Array{T, 2}</code> (multivariate Gaussian traits). For multivariate traits, each    column of <code>y</code> should be a sample. </li><li><code>x</code>: Genotype matrix (an <code>Array{T, 2}</code> or <code>SnpLinAlg</code>). For univariate   analysis, samples are rows of <code>x</code>. For multivariate analysis, samples are   columns of <code>x</code> (i.e. input <code>Transpose(x)</code> for <code>SnpLinAlg</code>)</li><li><code>z</code>: Matrix of non-genetic covariates of type <code>Array{T, 2}</code> or <code>Array{T, 1}</code>.   For univariate analysis, sample covariates are rows of <code>z</code>. For multivariate   analysis, sample covariates are columns of <code>z</code>. If this is not specified, an   intercept term will be included automatically. If <code>z</code> is specified, make sure   the first column (row) is all 1s to represent the intercept. </li></ul><p><strong>Optional Arguments:</strong></p><ul><li><code>path</code>: Different values of <code>k</code> that should be tested. One can input a vector of    <code>Int</code> (e.g. <code>path=[5, 10, 15, 20]</code>) or a range (default <code>path=1:20</code>).</li><li><code>q</code>: Number of cross validation folds. Larger means more accurate and more computationally   intensive. Should be larger 2 and smaller than 10. Default <code>q=5</code>.</li><li><code>d</code>: Distribution of phenotypes. Specify <code>Normal()</code> for quantitative traits,   <code>Bernoulli()</code> for binary traits, <code>Poisson()</code> or <code>NegativeBinomial()</code> for   count traits, and <code>MvNormal()</code> for multiple quantitative traits. </li><li><code>l</code>: A link function. The recommended link functions are <code>l=IdentityLink()</code> for   quantitative traits, <code>l=LogitLink()</code> for binary traits, <code>l=LogLink()</code> for Poisson   distribution, and <code>l=Loglink()</code> for NegativeBinomial distribution. For multivariate   analysis, the choice of link does not matter. </li><li><code>zkeep</code>: BitVector determining whether non-genetic covariates in <code>z</code> will be subject    to sparsity constraint. <code>zkeep[i] = true</code> means covariate <code>i</code> will NOT be projected.   Note covariates forced in the model are not subject to sparsity constraints in <code>path</code>. </li><li><code>est_r</code>: Symbol (<code>:MM</code>, <code>:Newton</code> or <code>:None</code>) to estimate nuisance parameters for negative binomial regression</li><li><code>group</code>: vector storing group membership for each predictor</li><li><code>weight</code>: vector storing vector of weights containing prior knowledge on each predictor</li><li><code>folds</code>: Vector that separates the sample into <code>q</code> disjoint subsets</li><li><code>debias</code>: Boolean indicating whether we should debias at each IHT step. Defaults <code>false</code></li><li><code>verbose</code>: Boolean indicating whether to print progress and mean squared error for   each <code>k</code> in <code>path</code>. Defaults <code>true</code></li><li><code>max_iter</code>: is the maximum IHT iteration for a model to converge. Defaults to 100 </li><li><code>min_iter</code>: is the minimum IHT iteration before checking for convergence. Defaults to 5.</li><li><code>init_beta</code>: Whether to initialize beta values to univariate regression values.    Currently only Gaussian traits can be initialized. Default <code>false</code>. </li><li><code>memory_efficient</code>: If <code>true,</code> it will cause ~1.1 times slow down but one only   needs to store the genotype matrix (requiring 2np bits for PLINK binary files   and <code>8np</code> bytes for other formats). If <code>memory_efficient=false</code>, one may potentially   store <code>t</code> sparse matrices of dimension <code>8nk</code> bytes where <code>k</code> are the cross validated   sparsity levels. </li></ul><p><strong>Output</strong></p><ul><li><code>mse</code>: A vector of mean-squared error for each <code>k</code> specified in <code>path</code>. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/cross_validation.jl#L1-L59">source</a></section></article><h2 id="Specifying-Groups-and-Weights"><a class="docs-heading-anchor" href="#Specifying-Groups-and-Weights">Specifying Groups and Weights</a><a id="Specifying-Groups-and-Weights-1"></a><a class="docs-heading-anchor-permalink" href="#Specifying-Groups-and-Weights" title="Permalink"></a></h2><p>When you have group and weight information, you input them as optional arguments in <code>fit_iht</code> and <code>cv_iht</code>. The weight vector is a vector of <code>Float64</code>, while the group vector is a vector of <code>Int</code>. For instance,</p><pre><code class="language-Julia hljs">    g = #import group vector
    w = #import weight vector
    ng = length(unique(g)) # specify number of non-zero groups
    result = fit_iht(y, x, z; J=ng, k=10, d=Normal(), l=IdentityLink(), group=g, weight=w)</code></pre><h2 id="Simulation-Utilities"><a class="docs-heading-anchor" href="#Simulation-Utilities">Simulation Utilities</a><a id="Simulation-Utilities-1"></a><a class="docs-heading-anchor-permalink" href="#Simulation-Utilities" title="Permalink"></a></h2><p>For complex simulations, please use <a href="https://github.com/OpenMendel/TraitSimulation.jl">TraitSimulation.jl</a>. </p><p>MendelIHT provides very naive simulation utilities, which were written before <a href="https://github.com/OpenMendel/TraitSimulation.jl">TraitSimulation.jl</a> was developed.</p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.simulate_random_snparray" href="#MendelIHT.simulate_random_snparray"><code>MendelIHT.simulate_random_snparray</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">simulate_random_snparray(s::String, n::Integer, p::Integer; 
    [mafs::Vector{Float64}], [min_ma::Integer])</code></pre><p>Creates a random SnpArray in the current directory without missing value,  where each SNP has ⫺5 (default) minor alleles. </p><p>Note: if supplied minor allele frequency is extremely small, it could take a long time for the simulation to generate samples where at least <code>min_ma</code> (defaults to 5) are present. </p><p><strong>Arguments:</strong></p><ul><li><code>s</code>: name of SnpArray that will be created in the current directory. To not   create file, use <code>undef</code>.</li><li><code>n</code>: number of samples</li><li><code>p</code>: number of SNPs</li></ul><p><strong>Optional Arguments:</strong></p><ul><li><code>mafs</code>: vector of desired minor allele freuqencies (uniform(0,0.5) by default)</li><li><code>min_ma</code>: the minimum number of minor alleles that must be present for each   SNP (defaults to 5)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/simulate_utilities.jl#L1-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.simulate_correlated_snparray" href="#MendelIHT.simulate_correlated_snparray"><code>MendelIHT.simulate_correlated_snparray</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">simulate_correlated_snparray(s, n, p; block_length, hap, prob)</code></pre><p>Simulates a SnpArray with correlation. SNPs are divided into blocks where each adjacent SNP is the same with probability prob. There are no correlation between blocks.</p><p><strong>Arguments:</strong></p><ul><li><code>n</code>: number of samples</li><li><code>p</code>: number of SNPs</li><li><code>s</code>: name of SnpArray that will be created (memory mapped) in the current directory. To not memory map, use <code>undef</code>.</li></ul><p><strong>Optional arguments:</strong></p><ul><li><code>block_length</code>: length of each LD block</li><li><code>hap</code>: number of haplotypes to simulate for each block</li><li><code>prob</code>: with probability <code>prob</code> an adjacent SNP would be the same. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/simulate_utilities.jl#L103-L118">source</a></section></article><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Simulating a SnpArray with <span>$n$</span> subjects and <span>$p$</span> SNPs requires up to <span>$2np$</span> bits of RAM. </p></div></div><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.simulate_random_response" href="#MendelIHT.simulate_random_response"><code>MendelIHT.simulate_random_response</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">simulate_random_response(x, k, d, l; kwargs...)</code></pre><p>This function simulates a random response (trait) vector <code>y</code>. When the  distribution <code>d</code> is from Poisson, Gamma, or Negative Binomial, we simulate  <code>β ∼ N(0, 0.3)</code> to roughly ensure the mean of response <code>y</code> doesn&#39;t become too large. For other distributions, we choose <code>β ∼ N(0, 1)</code>. </p><p><strong>Arguments</strong></p><ul><li><code>x</code>: Design matrix</li><li><code>k</code>: the true number of predictors. </li><li><code>d</code>: The distribution of the simulated trait (note <code>typeof(d) = UnionAll</code> but <code>typeof(d())</code> is an actual distribution: e.g. Normal)</li><li><code>l</code>: The link function. Input <code>canonicallink(d())</code> if you want to use the canonical link of <code>d</code>.</li></ul><p><strong>Optional arguments</strong></p><ul><li><code>r</code>: The number of success until stopping in negative binomial regression, defaults to 10</li><li><code>α</code>: Shape parameter of the gamma distribution, defaults to 1</li><li><code>Zu</code>: Effect of non-genetic covariates. <code>Zu</code> should have dimension <code>n × 1</code>. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/simulate_utilities.jl#L188-L206">source</a></section><section><div><pre><code class="nohighlight hljs">simulate_random_response(x, k, traits)</code></pre><p>Simulates a response matrix <code>Y</code> where each row is an independent multivariate Gaussian with length <code>trait</code>. There are <code>k</code> non-zero <code>β</code> over all traits. Each trait shares <code>overlap</code> causal SNPs. The covariance matrix <code>Σ</code> is positive definite and symmetric.</p><p><strong>Arguments</strong></p><ul><li><code>x</code>: Design matrix of dimension <code>n × p</code>. Each row is a sample. </li><li><code>k</code>: the total true number of causal SNPs (predictors)</li><li><code>traits</code>: Number of traits</li></ul><p><strong>Optional arguments</strong></p><ul><li><code>Zu</code>: Effect of non-genetic covariates. <code>Zu</code> should have dimension <code>n × traits</code>. </li><li><code>overlap</code>: Number of causal SNPs shared by all traits. Shared SNPs does not have the same effect size. </li></ul><p><strong>Outputs</strong></p><ul><li><code>Y</code>: Response matrix where each row is sampled from a multivariate normal with mean <code>μ[i] = X[i, :] * true_b</code> and variance <code>Σ</code></li><li><code>Σ</code>: the symmetric, positive definite covariance matrix used</li><li><code>true_b</code>: A sparse matrix containing true beta values.</li><li><code>correct_position</code>: Non-zero indices of <code>true_b</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/simulate_utilities.jl#L243-L265">source</a></section></article><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>For negative binomial and gamma, the link function must be LogLink. </p></div></div><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.make_bim_fam_files" href="#MendelIHT.make_bim_fam_files"><code>MendelIHT.make_bim_fam_files</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">make_bim_fam_files(x::SnpArray, y, name::String)</code></pre><p>Creates .bim and .bed files from a SnpArray. </p><p><strong>Arguments:</strong></p><ul><li><code>x</code>: A SnpArray (i.e. <code>.bed</code> file on the disk) for which you wish to create corresponding <code>.bim</code> and <code>.fam</code> files.</li><li><code>name</code>: string that should match the <code>.bed</code> file (Do not include <code>.bim</code> or <code>.fam</code> extensions in <code>name</code>).</li><li><code>y</code>: Trait vector that will go in to the 6th column of <code>.fam</code> file. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/simulate_utilities.jl#L350-L359">source</a></section></article><h2 id="Other-Useful-Functions"><a class="docs-heading-anchor" href="#Other-Useful-Functions">Other Useful Functions</a><a id="Other-Useful-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Other-Useful-Functions" title="Permalink"></a></h2><p>MendelIHT additionally provides useful utilities that may be of interest to a few advanced users. </p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.iht_run_many_models" href="#MendelIHT.iht_run_many_models"><code>MendelIHT.iht_run_many_models</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Runs IHT across many different model sizes specifed in <code>path</code> using the full design matrix. Same as <code>cv_iht</code> but <strong>DOES NOT</strong> validate in a holdout set, meaning that this will definitely induce overfitting as we increase model size. Use this if you want to quickly estimate a range of feasible model sizes before  engaging in full cross validation. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/cross_validation.jl#L225-L231">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.pve" href="#MendelIHT.pve"><code>MendelIHT.pve</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">pve(y, X, β; l = IdentityLink())</code></pre><p>Estimates phenotype&#39;s Proportion of Variance Explained (PVE) by typed genotypes  (i.e. chip heritability or SNP heritability).</p><p><strong>Model</strong></p><p>We compute <code>Var(ŷ) / Var(y)</code> where <code>y</code> is the raw phenotypes, <code>X</code> contains  all the genotypes, and <code>ŷ = Xβ</code> is the predicted (average) phenotype values from the statistical model β. Intercept is NOT included.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/c43c18e5a9619f7a5de4d9ca37910521e96b298a/src/pve.jl#L1-L11">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>parse_genotypes</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>convert_gt</code>. Check Documenter&#39;s build log for details.</p></div></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../FAQ/">« FAQ</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.25 on <span class="colophon-date" title="Sunday 13 August 2023 02:14">Sunday 13 August 2023</span>. Using Julia version 1.9.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
